src/tpm2-crypto.c

```diff
@@ -262,6 +265,182 @@ UsefulBuf read_ek_cert(void)
     return ret;
 }
 
+static UsefulBuf _cbor_cert_chain(UsefulBuf buf, size_t num, UsefulBufC *certs)
+{
+    QCBOREncodeContext ctx;
+    UsefulBufC enc;
+    QCBORError err;
+    QCBOREncode_Init(&ctx, buf);
+
+    QCBOREncode_OpenMap(&ctx);
+        QCBOREncode_OpenArrayInMap(&ctx, "certs");
+        for (int i = num - 1; i >= 0; i--)
+            QCBOREncode_AddBytes(&ctx, certs[i]);
+        QCBOREncode_CloseArray(&ctx);
+    QCBOREncode_CloseMap(&ctx);
+
+    err = QCBOREncode_Finish(&ctx, &enc);
+
+    if(err != QCBOR_SUCCESS) {
+        fprintf(stderr, "QCBOR error: %d\n", err);
+        return NULLUsefulBuf;
+    } else {
+        return UsefulBuf_Unconst(enc);
+    }
+}
+
+/*
+ * Authority Info Access and CA Issuers field is required by Windows 10 and
+ * newer, but not strictly required by TPM specification. This may render some
+ * TPMs unusable, but I haven't seen a TPM that has EK certificate without those
+ * fields.
+ */
+/* Based on X509_get1_ocsp() */
+static unsigned char *X509_get_ca_url(X509 *x)
+{
+    AUTHORITY_INFO_ACCESS *info;
+    unsigned char *ret = NULL;
+    int i, len;
+
+    info = X509_get_ext_d2i(x, NID_info_access, NULL, NULL);
+    if (!info)
+        return NULL;
+    for (i = 0; i < sk_ACCESS_DESCRIPTION_num(info); i++) {
+        ACCESS_DESCRIPTION *ad = sk_ACCESS_DESCRIPTION_value(info, i);
+        if (OBJ_obj2nid(ad->method) == NID_ad_ca_issuers) {
+            if (ad->location->type == GEN_URI) {
+                len = ASN1_STRING_to_UTF8(&ret, ad->location->d.uniformResourceIdentifier);
+                if (len < 0)
+                    ret = NULL;
+                break;
+            }
+        }
+    }
+    AUTHORITY_INFO_ACCESS_free(info);
+    return ret;
+}
+
+static X509 *UsefulBufC_to_X509(UsefulBufC ub)
+{
+    /* Pointer to buffer is modified by d2i_* functions, make a copy */
+    const unsigned char *tmp_ptr = ub.ptr;
+    return d2i_X509(NULL, &tmp_ptr, ub.len);
+}
+
+static size_t curl_cb(void *contents, size_t size, size_t nmemb, void *userp)
+{
+    size_t realsize = size * nmemb;
+    UsefulBuf *mem = (UsefulBuf *)userp;
+
+    char *ptr = realloc(mem->ptr, mem->len + realsize);
+    if(!ptr) {
+        /* out of memory! */
+        printf("not enough memory (realloc returned NULL)\n");
```

Comment:
I would use
```c
fprintf(stderr, "Not enough memory (...)\n");
```
And may be even use `strerror(errno)`

Reply:
Ack on `stderr`, but I don't think `strerror()` is useful in this case. It may possibly mix our error printed in English and whatever `errno` holds in language based on user's locale. Also, there really aren't many possibilities for `realloc()` to fail anyway.